### DISCUSSION TITLE
Inherit and modify access / visibility on fields from ObjectType in InputType

### DISCUSSION THREAD

**Discussion by theseyi** [MichalLytek/type-graphql#876] [Q&A] [answered]
What's the recommended way to inherit fields from an ObjectType into an InputType but modify the access to a subset of those fields.
For example
```ts
// This would be the @ObjectType, but inheritance between types with differing decorators is not supported
class User {
  @Field()
  firstName = '';
  @Field()
  lastName = '';
  @Field()
  fullName = this.firstName + ' ' + this.lastName;
}

// This would be the @InputType, but with the fullName field losing the field decorator so it is not user editable
class UserInput extends User {
    fullName = this.firstName + ' ' + this.lastName;
}
```

I'm aware you can reuse code between types but they both have to use the same decorators which does not meet the constraint here

From https://typegraphql.com/docs/inheritance.html#types-inheritance

> Note that both the subclass and the parent class must be decorated with the same type of decorator, like @ObjectType() in the example Person -> Student above. Mixing decorator types across parent and child classes is prohibited and might result in a schema building error, e.g. we can't decorate the subclass with @ObjectType() and the parent with @InputType().

**Comment by MichalLytek** [comment, marked as answer]
You should use the mixin-classes pattern.
This way you can declare the subparts of classes, like `firstName` and `lastName`, and then reuse them to compose input types and output types.

You can take a look at the example in the repo:
https://github.com/MichalLytek/type-graphql/tree/master/examples/mixin-classes

**Reply by theseyi** [reply]
Ok, I've used this pattern many times before for other tasks, was wondering (or hoping) there might be a more lightweight pattern perhaps using inheritance. But this is fine if it's the primary recommendation
